### PR TITLE
Update netconf_config.py to include option for datastore and improve error handling

### PR DESCRIPTION
**Pull request by userlerueda**
##### SUMMARY

This pull request addresses the problems described in #22191 to include the option for the user to select the datastore and improves error/exception handling.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

netconf_config

##### ANSIBLE VERSION

```
ansible 2.4.0 (devel 11138abc51) last updated 2017/04/01 12:09:33 (GMT -500)
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by userlerueda** [comment]
@wisotzky, @lpenz can you test with your respective systems ? I will test against tail-f nso.

**Comment by wisotzky** [comment]
Code has incompletely been merged:
- check line 158
- exception handling from ncclient removed

**Comment by userlerueda** [comment]
@wisotzky do you mean this one ?
```
$ git diff
diff --git a/lib/ansible/modules/network/netconf/netconf_config.py b/lib/ansible/modules/network/netconf/netconf_config.py
index c2636203a..e85ec9b84 100644
--- a/lib/ansible/modules/network/netconf/netconf_config.py
+++ b/lib/ansible/modules/network/netconf/netconf_config.py
@@ -155,7 +155,7 @@ import logging
 def netconf_edit_config(m, xml, commit, retkwargs, datastore):
     m.lock(target=datastore)
     try:
-        if ":candidate" in m.server_capabilities:
+        if datastore == "candidate":
             m.discard_changes()
         config_before = m.get_config(source=datastore)
         m.edit_config(target=datastore, config=xml)
```

**Comment by userlerueda** [comment]
@wisotzky let me know if I am missing anything, I tested against tail-f nso server and it is working, I tested auto, candidate, running, somethingelse and it properly responds with error codes.
```
TASK [Test netconf_config without specifying datastore] ****************************************************************************************************************************************************************
ok: [localhost]

TASK [Test netconf_config using auto datastore] ************************************************************************************************************************************************************************
ok: [localhost]

TASK [Test netconf_config using running datastore] *********************************************************************************************************************************************************************
ok: [localhost]

TASK [Test netconf_config using candidate datastore] *******************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": ":candidate is not supported by this netconf server"}
...ignoring

TASK [Test netconf_config using somethingelse datastore] ***************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "somethingelse datastore is not supported by this ansible module"}
...ignoring
```

**Comment by wisotzky** [comment]
Correct this is about the following lines:
```
-        if ":candidate" in m.server_capabilities:
+        if datastore == "candidate":
```

The discard operation (on candidate) should not be executed, if you want to enforce usage running. In my code proposal the method netconf_edit_config() does not longer check for the :candidate or :writeable-running capabilities - as all of these checks are already executed before executing this method.

The other change is the exception handling for following try/catch clause:
```
    try:
        changed = netconf_edit_config(
            m=m,
            xml=module.params['xml'],
            commit=True,
            datastore=datastore,
        )
+   except:
+       e = get_exception()
+       module.fail_json(msg='error editing configuration: ' + str(e))
        
    finally:
        m.close_session()
    module.exit_json(changed=changed, **retkwargs)
```

This exception handler should catch errors and generate an appropriate ansible error message reported by ncclient.

**Comment by userlerueda** [comment]
@wisotzky does it look good ? Please let us know and also if you can test it (ideally clone my repo to ensure you are using the exact same thing that has the patch).

**Comment by lpenz** [comment]
I've tested this with https://github.com/lpenz/ansible-testing-netconf, which uses openyuma, and it's ok.
Code is looking good too.

I'm just wondering if we shouldn't document a little better the fact that the selected datastore simply selects the mechanism used to activate the configuration. I mean, if you select datastore=candidate, the module sends the configuration to the candidate datastore *and then commits it*, which will have the effect of copying the candidate over the running datastore.

**Comment by userlerueda** [comment]
@lpenz what do you recommend we include in the comments or documentation ? I agree it could be dangerous if the user selects candidate and does not know what they are doing.

**Comment by wisotzky** [comment]
Default is auto, which in general is a good choice. People who know Netconf should be aware about the difference between running and candidate. Clearly you could clone the definition from Netconf RFC, but this reference exists anyhow already.

Sven


-------- Ursprüngliche Nachricht --------
Von: Luis Rueda <notifications@github.com>
Datum:04.04.2017 05:50 (GMT+01:00)
An: ansible/ansible <ansible@noreply.github.com>
Cc: "Wisotzky, Sven (Nokia - DE/Stuttgart)" <sven.wisotzky@nokia.com>, Mention <mention@noreply.github.com>
Betreff: Re: [ansible/ansible] Update netconf_config.py to include option for datastore and improve error handling (#23199)


@lpenz<https://github.com/lpenz> what do you recommend we include in the comments or documentation ? I agree it could be dangerous if the user selects candidate and does not know what they are doing.

**Comment by wisotzky** [comment]
Implementation of commit should not be just a copy. Usually it compares candidate with running and deploys the differences. This is important, as you don't want to have any service interruption or breakage in stats collection in services or protocols untouched.

Saying this, implementation of edit-config against running might be different dependent on the protocol stack used. There are implementations which process the edit-config request as single transactions - while others might do more imperative/ordered style config.

Unclear, if this really matters to playbook authors. Clearly you would always prefer transactions against no transactions - but it will always depend a bit on the actual implementation and users should be aware about it. Potentially this is one of the bad takes in the NETCONF RFC, as they provide some flexibility which leads to different behavior on different boxes.

/wiso


-------- Ursprüngliche Nachricht --------
Von: Leandro Lisboa Penz <notifications@github.com>
Datum:03.04.2017 23:25 (GMT+01:00)
An: ansible/ansible <ansible@noreply.github.com>
Cc: "Wisotzky, Sven (Nokia - DE/Stuttgart)" <sven.wisotzky@nokia.com>, Mention <mention@noreply.github.com>
Betreff: Re: [ansible/ansible] Update netconf_config.py to include option for datastore and improve error handling (#23199)


I've tested this with https://github.com/lpenz/ansible-testing-netconf, which uses openyuma, and it's ok.
Code is looking good too.

I'm just wondering if we shouldn't document a little better the fact that the selected datastore simply selects the mechanism used to activate the configuration. I mean, if you select datastore=candidate, the module sends the configuration to the candidate datastore and then commits it, which will have the effect of copying the candidate over the running datastore.

**Comment by lpenz** [comment]
Maybe be a bit more explicit by stating:

     - candidate, edit <candidate/> datastore and then commit
     - running, edit <running/> datastore directly

`auto` looks good already.

What do you think?

**Comment by wisotzky** [comment]
This should be okay. But maybe some words should be added on default operation, error option and test option. Seems the tool is using all defaults. Question is, if any of those is really used and if ansible playbook authors want to control it.

**Comment by userlerueda** [comment]
I have fixed conflict, squashed into single commit and updated with latest notes. Can you please double-check everything is good so we can send for merging.

**Comment by lpenz** [comment]
Tests done on ansible-testing-netconf
shipit

**Comment by userlerueda** [comment]
shipit

**Comment by wisotzky** [comment]
@userlerueda, it would rather make sense to update the parameter specs to make datastore a choice rather than a string:
```
datastore=dict(choices=['auto', 'candidate', 'running'], default='auto'),
```
Do you agree?

File: lib/ansible/modules/network/netconf/netconf_config.py [hunk 1]
```diff
@@ -181,6 +185,7 @@ def main():
             hostkey_verify=dict(type='bool', default=True),
             allow_agent=dict(type='bool', default=True),
             look_for_keys=dict(type='bool', default=True),
+            datastore=dict(type='str', default='auto'),
```
Comment (wisotzky) [inline lib/ansible/modules/network/netconf/netconf_config.py:188]: should use choice

File: lib/ansible/modules/network/netconf/netconf_config.py [hunk 2]
```diff
@@ -144,20 +152,16 @@
 import logging
 
 
-def netconf_edit_config(m, xml, commit, retkwargs):
-    if ":candidate" in m.server_capabilities:
-        datastore = 'candidate'
-    else:
-        datastore = 'running'
+def netconf_edit_config(m, xml, commit, retkwargs, datastore):
```
Comment (wisotzky) [inline lib/ansible/modules/network/netconf/netconf_config.py:155]: parameter retkwargs is not used;
should avoid using function parameters which are not implemented...

**Comment by userlerueda** [comment]
@wisotzky I think this pull request was already merged onto devel, I think any changes would have to come in as a new issue and therefore implemented using a new Pull Request, @lpenz can you confirm this ?

**Comment by lpenz** [comment]
Yup, it's merged: https://github.com/ansible/ansible/commit/753d30b64d50ac74f787294eb4db241b05441471

**Comment by wisotzky** [comment]
@userlerueda, can you raise an internal issue for this?

The improvement would be very limited, but follows better the ansible logic. Potentially the fix could just be embedded in one of the feature improvements I've suggested(#24220 or #24221).

**Pull request merged**